lib/Provider/Command/MessageSend.php

```diff
@@ -0,0 +1,137 @@
+
+use OCA\Mail\AppInfo\Application;
+use OCA\Mail\Db\LocalMessage;
+use OCA\Mail\Service\AccountService;
+use OCA\Mail\Service\Attachment\AttachmentService;
+use OCA\Mail\Service\OutboxService;
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\IConfig;
+use OCP\Mail\Provider\Exception\SendException;
+use OCP\Mail\Provider\IAddress;
+use OCP\Mail\Provider\IMessage;
+
+class MessageSend {
+
+	public function __construct(
+		protected IConfig $config,
+		protected ITimeFactory $time,
+		protected AccountService $accountService,
+		protected OutboxService $outboxService,
+		protected AttachmentService $attachmentService
+	) {
+	}
+
+	/**
+	 * Performs send operation
+	 *
+	 * @since 4.0.0
+	 *
+	 * @param string $userId system user id
+	 * @param string $serviceId mail account id
+	 * @param IMessage $message mail message object with all required parameters to send a message
+	 * @param array $options array of options reserved for future use
+	 *
+	 * @return LocalMessage
+	 *
+	 * @throws SendException on failure, check message for reason
+	 *
+	 */
+	public function perform(string $userId, string $serviceId, IMessage $message, array $options = []): LocalMessage {
+		// find user mail account details
+		try {
+			$account = $this->accountService->find($userId, (int)$serviceId);
+		} catch (\Throwable $e) {
+			throw new SendException('Error: occurred while retrieving mail account details', 0, $e);
+		}
+		// convert mail provider message to mail app message
+		$localMessage = new LocalMessage();
+		$localMessage->setType($localMessage::TYPE_OUTGOING);
+		$localMessage->setAccountId($account->getId());
+		$localMessage->setSubject($message->getSubject());
+		$localMessage->setBody($message->getBody());
+		// disabled due to issues caused by opening these messages in gui
+		//$localMessage->setEditorBody($message->getBody());
+		$localMessage->setHtml(true);
+		$localMessage->setSendAt($this->time->getTime());
+		// convert all mail provider attachments to local attachments
+		$attachments = [];
+		if (count($message->getAttachments()) > 0) {
+			// iterate attachments and save them
+			foreach ($message->getAttachments() as $entry) {
+				// determine if required parameters are set
+				if (empty($entry->getName()) || empty($entry->getType()) || empty($entry->getContents())) {
+					throw new SendException('Invalid Attachment Parameter: MUST contain values for Name, Type and Contents');
+				}
+				// convert mail provider attachment to mail app attachment
+				try {
+					$attachments[] = $this->attachmentService->addFileFromString(
+						$userId,
+						$entry->getName(),
+						$entry->getType(),
+						$entry->getContents()
+					)->jsonSerialize();
+				} catch (\Throwable $e) {
+					throw new SendException('Error: occurred while saving mail message attachment', 0, $e);
+				}
+			}
+		}
+		// determine if required To address is set
+		if (empty($message->getTo()) || empty($message->getTo()[0]->getAddress())) {
```

Comment:
Easiest way to fix the psalm warning about possibly null: Introduce a local variable for to.

A bit nicer: Move the validation to a function.

```php
	/**
	 * @param IAddress[] $addresses
	 * @throws SendException
	 */
	private function validateTo(array $addresses): void
	{
		if (count($addresses) === 0 || empty($addresses[0]->getAddress())) {
			throw new SendException('Invalid Message Parameter: MUST contain at least one TO address with a valid address');
		}
	}
```